src/c_string.rs

```diff
@@ -0,0 +1,597 @@
+    ///     - `ptr` must be non-nul even for a zero-length `CStr`.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use core::ffi::{c_char, CStr};
+    /// use heapless::CString;
+    ///
+    /// const HELLO_PTR: *const c_char = {
+    ///     const BYTES: &[u8] = b"Hello, world!\0";
+    ///     BYTES.as_ptr().cast()
+    /// };
+    ///
+    /// let copied = unsafe { CString::<14>::from_raw(HELLO_PTR) }.unwrap();
+    ///
+    /// assert_eq!(copied.to_str(), Ok("Hello, world!"));
+    /// ```
+    pub unsafe fn from_raw(ptr: *const c_char) -> Result<Self, ExtendError> {
+        // SAFETY: The given pointer to a string is assumed to be nul-terminated.
+        Self::from_bytes_with_nul(unsafe { CStr::from_ptr(ptr).to_bytes_with_nul() })
+    }
+
+    /// Converts the [`CString`] to a [`CStr`] slice.
+    pub fn as_c_str(&self) -> &CStr {
+        debug_assert!(CStr::from_bytes_with_nul(&self.inner).is_ok());
+
+        unsafe { CStr::from_bytes_with_nul_unchecked(&self.inner) }
+    }
+
+    /// Calculates the length of `self.inner` would have if it appended `bytes`.
+    fn capacity_with_bytes(&self, bytes: &[u8]) -> Option<usize> {
+        match bytes.last() {
+            None => None,
+            Some(0) if bytes.len() < 2 => None,
+            Some(0) => {
+                // `bytes` is nul-terminated and so is `self.inner`.
+                // Adding up both would account for 2 nul bytes when only a single byte
+                // would end up in the resulting CString.
+                Some(self.inner.len() + bytes.len() - 1)
+            }
+            Some(_) => {
+                // No terminating nul byte in `bytes` but there's one in
+                // `self.inner`, so the math lines up nicely.
+                //
+                // In the case that `bytes` has a nul byte anywhere else, we would
+                // error after `memchr` is called. So there's no problem.
+                Some(self.inner.len() + bytes.len())
+            }
+        }
+    }
+
+    /// Extends the [`CString`] with the given bytes.
+    ///
+    /// This function fails if the [`CString`] would not have enough capacity to append the bytes or
+    /// if the bytes contain an interior nul byte.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use heapless::CString;
+    ///
+    /// let mut c_string = CString::<10>::new();
+    ///
+    /// c_string.extend_from_bytes(b"hey").unwrap();
+    /// c_string.extend_from_bytes(b" there\0").unwrap();
+    ///
+    /// assert_eq!(c_string.to_str(), Ok("hey there"));
+    /// ```
+    pub fn extend_from_bytes(&mut self, bytes: &[u8]) -> Result<(), ExtendError> {
+        let Some(capacity) = self.capacity_with_bytes(bytes) else {
+            return Ok(());
+        };
+
+        if capacity > N {
+            // Cannot store these bytes due to an insufficient capacity.
+            return Err(CapacityError.into());
+        }
+
+        match memchr(0, bytes) {
```

Comment:
I was wondering if we should use `CStr::from_bytes_with_nul` here directly (it uses the `memchr` crate)

```rust
    match CStr::from_bytes_with_nul(bytes) {
        Ok(_cstr) => {
            unsafe { self.extend_from_bytes_unchecked(bytes) }?;
            Ok(())
        },
        Err(FromBytesWithNulError::InteriorNul { position }) => {
            Err(ExtendError::InteriorNulByte(position))
        },
        Err(FromBytesWithNulError::NotNulTerminated) => {
            unsafe {
                self.extend_from_bytes_unchecked(bytes).unwrap();
                self.inner.push_unchecked(0);
            };
            Ok(())
        }
    }
}
```

Reply:
Makes sense.